oeps/oep-0045-arch-ops-and-config.rst

```diff
@@ -0,0 +1,152 @@
+
+
+**Clear System Dependency Documentation**
+
+For operators who don’t want to run in Docker containers the Dockerfile acts as documentation for the system dependencies they will need to recreate.  Everything an operator needs to know about how to set up their environment is all in one place rather than distributed across multiple overlapping and overriding files.
+
+Configuration
+*************
+ 
+**Django settings**
+
+Having a single artifact that runs with different configurations increases stability by improving development parity with other deployment environments.  edX applications already support configuration overrides via a yaml file for production environments, but development and test environments tend to configure the application using different code paths via a settings/devstack.py or settings/test.py file.
+
+Additionally it is not clear which settings are required to be overridden and which settings have values that may technically work but are inappropriate for production systems.  To alleviate these issue edX django applications will adopt the following settings structure:
+
+.. code-block:: text
+
+  settings
+  ├── __init__.py
+  ├── required.py
+  └── defaults.py
+
+
+*  ``__init__.py`` - Sourcing our config from this file within the settings directory takes advantage of django defaults and means that settings will be picked up automatically without needing to specify ``--settings`` anywhere.  This entry point would import ``required.py``, ``defaults.py``, and the code to override both from a config file.
+*  ``required.py`` - all settings which are required to run and do not have a reasonable production-ready default.  EG LMS_BASE_URL which will be different per environment.
+*  ``defaults.py`` - all settings other settings which will have production-ready defaults
+
+The settings defined in ``required.py`` and ``defaults.py`` files are mutually exclusive, representing all application specific settings as well as installed library settings whose values either must be provided or whose defaults are not considered production-ready.
+
+``required.py`` variables will all be initialized to ``NONE`` and the application will not start unless they are set.  This allows operators to fail fast rather than finding out about an unset value when users exercise those breaking codepaths. Application developers are encouraged to keep the list of required settings to a minimum.
+
+This new settings structure obviates the need for any other python files in the settings directory (such as ``devstack.py``, ``test.py``, etc).  The values currently set in those files should be moved to a corresponding ``devstack.yml``, ``test.yml``, etc in the same settings directory.  This gives developers and operators more consistency across environments since the same code paths are being executed with different values.
```

Comment:
Since this structure is mapped to the settings structure of the plugins for edx-platform, there could be a note about the path of upgrade for plugin maintainers. Perhaps keeping the old structure as a backwards compatibility layer for one open-release.

Reply:
Agreed.  We should certainly have a plan for this.  I would consider that part of the implementation strategy.  I would like to capture that plan in the OEP but separate the conversations.

Let's move this topic over to this follow-on PR https://github.com/edx/open-edx-proposals/pull/144